content/posts/mirrorbits-cdn.md

```diff
@@ -0,0 +1,392 @@
+
+## Enter Mirrorbits
+
+Luckily, I wasn't alone. Many years before, another FLOSS project had encountered the same problem. VideoLAN, creators of the fantastic VLC Media Player, had the same issue. So their talented developers created a solution: Mirrorbits.
+
+[Mirrorbits](https://github.com/etix/mirrorbits) is a fairly straightforard Go program with a single goal: provide a way to distribute requests from a single central repository to multiple georedundant repositories, based on the client's GeoIP information. It seemed to fit the bill exactly.
+
+But there was a problem: documentation on how to actually run Mirrorbits was sparse, so it took quite a bit of trial-and-error to determine how to use it. I hope the following will help avoid this trouble for anyone else.
+
+## File layout
+
+The first thing to consider is the layout of the files. In Jellyfin's case, our repository is large and sprawling, constituting many different components including the server, clients, plugins, and various secondary files. Mirrorbits requires everything to be housed under one directory, and we needed a way to synchronize this whole directory easily. We also had a problem of our "archives", old stable releases that we did not need or want sychronized to all of our mirror servers wasting space.
+
+The solution we came up with was the following directory structure:
+
+```
+/srv/repository
+               /mirror
+                      /releases
+                               /client
+                               /server
+                               /plugin
+                               etc.
+                      /debian
+                             /dists
+                             /pool
+                      /ubuntu
+                             /dists
+                             /pool
+                      /archive
+                      etc.
+               /releases --symlink--> /mirror/releases
+               /debian   --symlink--> /mirror/debian
+               etc.
+```
+
+In effect, everything is under a `/mirror` directory, with external symlinks to provide the root links we wanted.
+
+## Additional VPSes and Synchronization
+
+The next step after crafting a usable file layout was to create some additional VPSes and determine how to synchronize the files.
+
+Since our origin server is in Toronto, we wanted to ensure we had wide geographic coverage. And because of how Mirrorbits works (we'll get to that later), we also wanted a dedicated CDN for the same region as the origin server. Thus, we created 4 additional VPSes:
+
+* `tor1.mirror.jellyfin.org`: Toronto, Ontario, Canada for North/South America East
+* `sfo1.mirror.jellyfin.org`: San Francisco, California, USA for North/South America West
+* `fra1.mirror.jellyfin.org`: Frankfurt, Germany (not France as commonly assumed!) for Europe and Africa
+* `sgp1.mirror.jellyfin.org`: Singapore, Malaysia for Asia and Pacific
```

Comment:
Isn't this just Singapore?

Reply:
Haha yes it is, feels right to type though!